### PR TITLE
feat(clients): plugin reflex hooks (session digest, announce nudges)

### LINKED ISSUE #3131: Task: plugin reflex hooks — SessionStart broadcast/memory injection, announce and report reminders
Parent initiative: #3128
Depends-on: #3130

## Target

- New: `clients/claude-code-plugin/hooks/hooks.json` + hook scripts under `clients/claude-code-plugin/bin/`
- Modify: the plugin README (hook behavior + the scoped-matcher note)

## Context

Prompts and skills are advisory; hooks are deterministic — the Claude Code harness executes them, no model initiative required. This task adds the client-side reflex enforcement to the `meho` plugin: read-before-start happens *before the first prompt*, and announce-before-mutate / report-on-completion get non-blocking reminders. (The server-side counterpart — dispatch-time advisory + opt-in gate — is a sibling task; hooks are the only layer that can inject context at session start.)

## Current state

The plugin (once its skeleton task lands) ships skills and MCP config only. Whether an agent reads `broadcast_recent`, announces intent, or writes learnings back depends entirely on the model following advisory text.

## Desired state

Three hooks in `hooks/hooks.json`, all **fail-open** (missing `meho` CLI, no auth, no VPN, timeout ⇒ silent no-op, exit 0, session never breaks):

1. **`SessionStart`** — runs a script that fetches a compact digest (recent tenant broadcast window + scoped memory/knowledge recall) via the `meho` CLI and prints it to stdout, which Claude Code injects into the session as context. Read-before-start stops being a discipline the model must remember.
2. **`PreToolUse`** matched on the plugin's `call_operation` tool — if this session has not announced intent yet, emit a one-time, non-blocking reminder naming `broadcast_announce`. **Matcher gotcha:** tools from a plugin-bundled MCP server get a *scoped* name — `mcp__plugin_<plugin-name>_<server-name>__<tool>` — so the matcher must target that form (with a `.*` suffix where a server-wide match is wanted); a bare `mcp__meho__call_operation` never fires. Session announce-state via a marker file written by a `PostToolUse` hook on the announce tool.
3. **`Stop`** — if the session invoked `call_operation` but never announced/reported, emit a one-line report-on-completion reminder.

Warn-only throughout: no hook blocks a tool call — enforcement with teeth is the server-side gate's job.

## Acceptance criteria

- [ ] `SessionStart` context injection works: with a logged-in `meho` CLI on VPN, a new session's context contains the digest (verifiable via the session transcript); without CLI/auth/VPN the hook exits 0 in <2s and injects nothing
- [ ] Every hook entry sets an explicit `timeout`; scripts exit 0 on all failure paths (demonstrated by a test run with the CLI absent)
- [ ] The `PreToolUse` matcher uses the plugin-scoped tool-name form and demonstrably fires (transcript or debug output in the PR); the scoped-name formula is documented in the plugin README
- [ ] The announce reminder appears at most once per session (marker-file dedupe)
- [ ] No hook uses a blocking decision (no deny) — grep of `hooks.json` shows warn/context output only

## Out of scope

- Server-side nudges or gating (sibling task)
- Claude Desktop (no hook mechanism there — Desktop is covered by descriptions + server advisory)
- Auto-installing or configuring the `meho` CLI

## Verification steps

```bash
jq . clients/claude-code-plugin/hooks/hooks.json           # valid, timeouts present
PATH=/nonexistent clients/claude-code-plugin/bin/session-start.sh; echo $?   # 0, fast, silent
# manual: fresh session with plugin installed → digest visible; call_operation without announce → single reminder
```

## References

- Hooks: <https://code.claude.com/docs/en/hooks.md> — events (SessionStart / PreToolUse / PostToolUse / Stop), matcher semantics (regex when non-alphanumeric, MCP tool names `mcp__<server>__<tool>`, **plugin-scoped** `mcp__plugin_<plugin>_<server>__<tool>`), stdout-as-context for SessionStart, `${CLAUDE_PLUGIN_ROOT}`
- Server-assembled tenant preamble already *states* the broadcast discipline (#2546 seeded it) — these hooks are the client-side enforcement of the same rules; don't duplicate the preamble text, enforce it
- [AI engineering best practices: Safety defaults](../blob/main/.claude/skills/implement-issue/ai_engineering_best_practices.md#safety-defaults), [Agent-loop boundary](../blob/main/.claude/skills/implement-issue/ai_engineering_best_practices.md#agent-loop-boundary); [DevOps best practices: Pipeline axioms](../blob/main/.claude/skills/implement-issue/devops_best_practices.md#pipeline-axioms) (fail-open discipline for non-critical automation)

### PR DESCRIPTION
**Pull request by damir-topic**
## Summary

- Adds three **fail-open, warn-only** Claude Code reflex hooks to the `meho` plugin (`clients/claude-code-plugin/hooks/hooks.json` + scripts under `bin/`), turning the advisory Layer-2 discipline into deterministic, harness-run behavior.
- **SessionStart** injects a compact digest (recent tenant activity as the durable broadcast window, scoped memory, recent knowledge) via the `meho` CLI's stdout — read-before-start stops being a discipline the model must remember.
- **PreToolUse** on `call_operation` emits a one-time announce reminder naming `broadcast_announce`; **PostToolUse** on `broadcast_announce` records the announce; **Stop** reminds to report when `call_operation` ran but nothing was announced/reported.
- No hook ever denies or blocks — enforcement with teeth is the server-side announce gate's job (sibling #3133). Every failure path (missing CLI / expired login / no VPN / timeout) exits 0 silently.

## Design notes

- **Broadcast window source.** The live feed (`meho status --watch`) is an SSE stream, unusable from a hook that must return promptly. Every MEHO op writes an audit row and emits a broadcast event, so `meho audit recent` is the bounded, non-streaming read of the same window.
- **Scoped-matcher gotcha.** A plugin-bundled MCP server's tools are named `mcp__plugin_<plugin>_<server>__<tool>`. Both the plugin and its server are `meho`, so the matchers target `mcp__plugin_meho_meho__call_operation` / `mcp__plugin_meho_meho__broadcast_announce`; a bare `mcp__meho__...` never fires. The formula is documented in the plugin README.
- **Dedupe.** Session state (`used` / `announced` / `reminded` / `stop_reminded`) lives in per-`session_id` marker files under `${TMPDIR:-/tmp}/meho-plugin-hooks`, so the PreToolUse announce reminder and the Stop report reminder each fire at most once per session.
- **Non-blocking channels.** PreToolUse/Stop return the nudge as `hookSpecificOutput.additionalContext` (no `permissionDecision`), and neither ever exits 2. Note the real Stop semantics: `additionalContext` on a Stop hook does **not** end the turn — per the hooks docs it *continues* the conversation once (so report-on-completion gets a chance to fire), under the same loop protections as `decision: block` (the `stop_hook_active` input + the 8-consecutive-continuation cap). The Stop hook bounds that to a single nudge: it no-ops when `stop_hook_active` is `true` and writes a once-per-session `.stop_reminded` marker, then lets the session stop. SessionStart uses plain stdout, which Claude Code injects as context.
- Reminders **enforce** the broadcast discipline the server-assembled tenant preamble already states; they don't restate it.

## Test plan

- [x] `jq . clients/claude-code-plugin/hooks/hooks.json` — valid; all 4 hook entries carry an explicit `timeout` (12/5/5/5).
- [x] `PATH=/nonexistent clients/claude-code-plugin/bin/session-start.sh; echo $?` — exit `0`, silent, sub-second (CLI-absent fast path via a `command -v meho` builtin guard).
- [x] SessionStart happy path (stub `meho` on PATH) prints the three-section digest to stdout; when `meho status` errors (no auth/VPN) it prints nothing and exits 0.
- [x] PreToolUse on `call_operation` (not yet announced) emits valid-JSON `additionalContext` naming `broadcast_announce`; a second call is silent (marker-file dedupe → at most once per session).
- [x] Announce-first (PostToolUse `broadcast_announce`) suppresses the PreToolUse reminder.
- [x] Stop emits a report reminder only when `call_operation` was used but nothing was announced/reported; silent otherwise.
- [x] Stop is bounded to a single nudge (B1): `stop_hook_active=true` → silent exit 0 (no marker consumed); a second Stop in the same session after the first reminder → silent (`.stop_reminded` marker); the first used-but-not-announced Stop → exactly one valid-JSON reminder.
- [x] Empty / garbage stdin and an unwritable tmp dir all exit 0 (fail-open).
- [x] `shellcheck -s sh bin/*.sh` — clean. Scripts committed `100755`.
- [x] No blocking decision anywhere: grep of `hooks.json` + scripts shows no `deny` / `permissionDecision:deny` / `exit 2` / `escalate` — warn/context output only.

## Out of scope

- Server-side nudges / gating (sibling #3133, merged).
- Claude Desktop (no hook mechanism; covered by descriptions + server advisory).
- Auto-installing or configuring the `meho` CLI.

---

## Follow-up (auto-implement-issue, iteration 1, 2026-08-26)

Addressed review finding from `pr-3142-iter-1`:

- **B1** `67f4b196` — `bin/stop-report.sh` emitted `additionalContext` on Stop unconditionally, which per the hooks docs *continues* the turn (not a passive no-op), re-nagging the used-but-not-announced path toward the 8-continuation cap. Now reads `stop_hook_active` and no-ops when true, writes a once-per-session `.stop_reminded` marker so the reminder fires at most once, and the header comment + README Stop row / marker note now state the real continuation semantics. Re-ran `shellcheck -s sh bin/*.sh` (clean) plus extended behavioral checks (`stop_hook_active=true` → silent; second Stop in-session → silent; first used-but-not-announced Stop → single reminder).

Closes #3131

### CODE REVIEW

File: clients/claude-code-plugin/bin/stop-report.sh [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+payload="$(cat 2>/dev/null || true)"
+sid="$(printf '%s' "$payload" | sed -n 's/.*"session_id"[[:space:]]*:[[:space:]]*"\([^"]*\)".*/\1/p' | head -n 1)"
+sid="$(printf '%s' "$sid" | tr -cd 'A-Za-z0-9._-')"
+[ -n "$sid" ] || sid="nosession"
+
+used_marker="${state_dir}/${sid}.used"
+announced_marker="${state_dir}/${sid}.announced"
+
+if [ -e "$used_marker" ] && [ ! -e "$announced_marker" ]; then
```
Comment (damir-topic) [inline clients/claude-code-plugin/bin/stop-report.sh]: **[🛑 B1]** — Stop `additionalContext` continues the turn; without a `stop_hook_active` guard this re-nags up to the 8-continuation cap

**What:** This block emits `hookSpecificOutput.additionalContext` on the Stop event whenever `call_operation` ran without an announce. Per the current Claude Code hooks docs, `additionalContext` on Stop *"keeps the conversation going through the same loop protections as `decision: block`, namely the `stop_hook_active` input and the 8-consecutive-continuation cap"* — it does **not** let the turn end. The header comment (L7–L10) states the opposite ("exits 0, so the session stops normally"), so the implementation rests on an incorrect reading of the semantics. Because the hook never reads `stop_hook_active` and keeps no once-per-session marker (unlike the PreToolUse reminder's `.reminded`), a session that used `call_operation` without announcing — the common read-only path — is force-continued on Stop and re-nagged on every subsequent Stop until it either announces or the harness's 8-continuation cap fires.

**Why it matters:** This contradicts the task's defining contract (warn-only, advisory, "the session never breaks") and the fail-open discipline the PR is built around. Reads dominate `call_operation` usage, so most sessions hit `.used && !.announced` at end-of-turn and pay up to 8 forced model turns of nagging — wasted tokens and a blocked clean stop, not the passive one-line reminder the code intends.

**Required change:** Extract `stop_hook_active` from the hook's JSON stdin (same `sed`-extract pattern already used for `session_id`) and no-op when it is `true`; optionally also write a `${sid}.stop_reminded` marker and skip when it exists. Either bounds the report nudge to exactly one continuation, then lets the session stop. Also correct the L7–L10 comment (and the README's "exit 0 — never exit 2, which would block the stop" framing) to state that `additionalContext` on Stop *does* continue the turn — that is fine and intended for report-on-completion, but it is not a passive no-op.

**Source:** https://code.claude.com/docs/en/hooks.md — Stop decision control: "`hookSpecificOutput.additionalContext` — Non-error feedback for Claude. The conversation continues so Claude can act on it …" and "It keeps the conversation going through the same loop protections as `decision: block`, namely the `stop_hook_active` input and the 8-consecutive-continuation cap."

**Comment by damir-topic** [comment]
> **DEGRADED MODE — machine review identity unavailable.** This
> verdict is advisory prose, **not** a GitHub review;
> `reviewDecision` is unchanged and branch protection is not
> satisfied by this comment. (The `meho-review-app` 1Password item is
> not provisioned in any accessible vault, so a formal
> `--request-changes` submission would 422 as a self-review.)

## /auto-review-pr summary (iteration 1)




**Anchor task:** #3131 — plugin reflex hooks (SessionStart digest + announce/report reminders)
**Surface area:** devops (shell hooks + plugin JSON) + docs
**Best-practices packs audited:** devops (pipeline axioms / fail-open discipline), ai (agent-loop boundary)
**Verdict:** REQUEST_CHANGES
**Head SHA reviewed:** `bb242ce9`

### Acceptance criteria
- [x] SessionStart context injection via **plain stdout** — correct channel (docs: "Claude Code adds plain-text stdout as context" for SessionStart). Digest assembles the three sections on the happy path; status-gate short-circuits with no sub-read leak; empty sub-reads emit nothing. Verified with a stub `meho`.
- [x] Every hook entry sets an explicit `timeout` (12/5/5/5); all scripts exit 0 on every failure path — CLI absent, empty stdin, garbage stdin, closed stdin, unwritable `TMPDIR`. Verified empirically.
- [x] `PreToolUse` matcher uses the plugin-scoped form `mcp__plugin_meho_meho__call_operation`; plugin name (`.claude-plugin/plugin.json`) and server key (`.mcp.json`) are **both** `meho`, so the scoped name is correct on both segments. Formula documented in README. `PostToolUse` matcher likewise correct.
- [x] Announce (PreToolUse) reminder fires **at most once** per session via the `.reminded` marker; a prior `broadcast_announce` (PostToolUse → `.announced`) suppresses it. Verified.
- [x] No hook uses `deny` / `permissionDecision:"deny"` / `exit 2` — grep-clean; warn/context output only.
- [❌] **Report-on-completion (Stop) reminder is advisory / non-blocking** — see 🛑 B1. The Stop hook's `additionalContext` continues the turn (docs-confirmed) and, lacking a `stop_hook_active` guard, can force up to 8 continuations on the common `call_operation`-without-announce path.

### Project-board alignment
| Issue | Status | Link | Auto-close | Delivered | Action |
|---|---|---|---|---|---|
| #3131 | Open (task) | Closes | ✅ | ⚠️ (one AC not met — B1) | close on merge once B1 fixed |
| #3128 | Open (parent initiative) | Refs (via task) | manual | partial (wave 3) | stays open |
| #3130 | Closed (skeleton) | dependency | — | ✅ (plugin skeleton on main) | — |
| #3133 | Closed (server-side gate) | sibling / out-of-scope | — | ✅ | — |

No stealth closures; the diff touches only the six declared plugin files.

### Predicted vs. actual diff
- ✅ `hooks/hooks.json`, `bin/session-start.sh`, `bin/pre-call-operation.sh`, `bin/post-announce.sh`, `bin/stop-report.sh`, `README.md` — all expected and present.
- ⚠️ `bin/stop-report.sh` — expected but wrong (B1).
- ❓ None unexpected.
- 🚫 None undisclosed.
- 🕳️ None missing. `bin/meho-mcp-remote`, `.claude-plugin/plugin.json`, `.mcp.json`, `.claude-plugin/marketplace.json` correctly **untouched** (no regression to the merged pinned-static-OAuth shim / manifest).

### CLI-verb reality (digest calls the real Go CLI)
- `meho status` ✅ (`cli/internal/cmd/status.go`; all failure paths → stderr, success → stdout, so the `[ -n "$status_out" ]` gate is a correct fail-open gate).
- `meho audit recent --limit N` ✅ (`cli/internal/cmd/audit/recent.go`, `Args: NoArgs`, `--limit` present).
- `meho list --limit N` ✅ — memory list is registered **at root** (`memory.NewListCmd()` in `root.go`), so `meho list` is the memory verb; `--limit` present.
- `meho kb list --limit N` ✅ (`cli/internal/cmd/kb/list.go`, `--limit` present).

### Integration verification
- [x] `jq . hooks.json` — valid; 4 entries, all with explicit `timeout`.
- [x] `shellcheck -s sh bin/*.sh` — clean; scripts committed `100755`.
- [x] Fail-open matrix (CLI absent / empty / garbage / closed stdin / unwritable tmp) — every path exit 0.
- [x] Dedupe + suppression semantics (PreToolUse once-only; announce-first suppresses; Stop emits only on used-and-not-announced; silent when never used) — all verified with marker-file harness.
- [x] Digest assembly (happy / status-fails-gate / empty-sub-reads) — verified with stub `meho`.
- [x] Hook-output JSON shapes verified against current Claude Code hooks docs (fetched fresh) — SessionStart plain stdout ✅; PreToolUse/PostToolUse/Stop `additionalContext` all documented + model-visible ✅ (but Stop's continuation semantics drive B1).
- [n/a] ruff / mypy / pytest / frontend — no Python/Go/TS touched. `CLI API snapshot freshness` CI check is green (no CLI change).

### Findings (inline — merge blocker per team policy)
- 🛑 **B1** (`bin/stop-report.sh:27`) — Stop `additionalContext` continues the turn (docs), and with no `stop_hook_active` guard / once-marker it re-nags up to the 8-continuation cap on the common read-without-announce path; the header comment (L7-10) claims the opposite. **Required change:** read `stop_hook_active` from stdin and no-op when `true` (and/or write a `${sid}.stop_reminded` marker), bounding the nudge to one continuation, then correct the L7-10 comment + the README's "exit 0 … would block the stop" framing.

### Adjacent improvements (non-blocking, not posted inline)
- **PreToolUse `additionalContext` lands *next to the tool result*** (docs) — i.e. after the first `call_operation` executes, not before it. Inherent to the warn-only channel (a true pre-execution gate would need `permissionDecision`, which the task forbids); the nudge still serves subsequent calls + completion. Worth a one-line comment in `pre-call-operation.sh` so a future reader doesn't expect a pre-mutation block.
- **`session-start.sh` status call is unbounded when neither `timeout` nor `gtimeout` is on PATH** (stock macOS without coreutils). Fail-open still holds via the 12s hook `timeout`, but AC1's "<2s" for the no-VPN path is not guaranteed there. The Go client uses `http.DefaultClient` (no timeout), so consider an explicit short client-side deadline, or document that the 2s bound requires a `timeout` binary.
- **Marker files are never cleaned up** — they accumulate under `${TMPDIR}/meho-plugin-hooks` (no `SessionEnd` hook to `rm` them). Tiny + in ephemeral tmp, so benign; a `SessionEnd` sweep would be tidy.
- **`nosession` fallback bucket collides across concurrent sessions** whose payloads lack a parseable `session_id`. Graceful degradation (worst case: one reminder is suppressed); real `session_id` is always supplied, so this is edge-case only.
- **No automated (bats) tests** for the shell hooks — manual verification is documented in the PR body and was reproduced here; a bats harness would lock the fail-open/dedupe contract against regressions.

### Docs
- [x] README "Reflex hooks" section is thorough and accurate on the scoped-matcher formula, fail-open contract, and digest source. One wording fix needed (part of B1): the Stop "exit 0 — never exit 2, which would block the stop" line implies exit-0 lets the session stop, but `additionalContext` on Stop continues the turn.
- [n/a] No `docs/codebase/` plugin doc exists; the plugin is documented via its own README (the convention established by skeleton #3130). Not a gap.
- [n/a] CHANGELOG correctly **not** touched — the CHANGELOG's declared scope is the image / chart / CLI, and the plugin (incl. its introduction in #3130) is out of that scope.

### What worked well (review-body only — never posted as inline comments)
- The scoped-matcher gotcha is handled correctly on **both** name segments and clearly documented — the exact failure mode the task warned about is avoided.
- Fail-open is genuinely airtight: builtin `command -v meho` fast-path, stderr-only CLI errors keeping the stdout gate honest, and exit-0 on every degenerate stdin/tmp condition — all verified empirically, not just claimed.
- Correct instinct on channels: plain stdout for SessionStart, `additionalContext` for the reminder events (the only model-visible channels for each), matching the fresh docs.

**Comment by damir-topic** [comment]
> **DEGRADED MODE — machine review identity unavailable.** This
> verdict is advisory prose, **not** a GitHub review;
> `reviewDecision` is unchanged and branch protection is not
> satisfied by this comment.

## /auto-review-pr summary (iteration 2)




**Anchor task:** #3131 — plugin reflex hooks (SessionStart digest, announce/report nudges)
**Surface area:** devops, docs
**Best-practices packs audited:** devops, ai
**Verdict:** APPROVE
**Head SHA reviewed:** `67f4b196`

Iteration 1 raised a single blocker (B1); this iteration verifies the fix
commit `67f4b196` and re-reviews the full current diff. **B1 is resolved.**

### B1 verification (the iter-1 blocker)
The Stop hook's `additionalContext` *continues* the turn rather than being a
passive no-op, so the iter-1 code re-nagged the common used-but-not-announced
path toward the 8-continuation cap. The fix adds two independent bounds and
corrects the docs, all verified here against the head SHA:

- **`stop_hook_active` guard** — reads the field from hook stdin and silently
  `exit 0` when `true`, so continuations never stack. Behaviorally confirmed:
  `stop_hook_active=true` (compact *and* pretty-printed multi-line JSON) →
  silent exit 0, no marker consumed.
- **Robust extraction** — the `sed` literal-`true` match (deliberately avoiding
  `\|` alternation for BSD/GNU portability) is false-positive-safe: a payload
  with `"stop_hook_active":false,"other_flag":true` still reminds (does not
  falsely bail), and `{...:true}` / `{...: true}` / `{...  :  true}` all
  extract `true`. `stop_hook_active` is a documented JSON boolean, so the
  literal match is correct.
- **Once-per-session `.stop_reminded` marker** — distinct from the PreToolUse
  `.reminded` marker; emits at most one reminder per session. Confirmed:
  first used-and-not-announced Stop → exactly one valid-JSON reminder + marker
  written; second Stop same session → silent.
- **No wrong-session leak** — markers are keyed by `session_id` (same `sed`
  extraction across all four scripts) and sanitized with `tr -cd
  'A-Za-z0-9._-'` (path-traversal-safe: `../../etc/evil` collapses to a flat
  name, no directory escape). Two concurrent sessions remind independently and
  each suppresses only its own second nudge.
- **Docs corrected** — the `stop-report.sh` header comment (L7–19) and the
  README Stop row + marker note now state the real "continues the turn once,
  under the same loop protections as `decision: block`" semantics; the PR body
  follow-up section documents the fix accurately.

### Acceptance criteria (#3131)
- [x] SessionStart injection + fail-open <2s — `command -v meho` fast path, 2s-bounded `status` gate, digest emitted only when a section has content. (See adjacent note on the unbounded case when neither `timeout` nor `gtimeout` is on PATH — fail-open still holds via the 12s hook timeout.)
- [x] Every hook sets an explicit `timeout` (12/5/5/5, jq-verified); scripts exit 0 on all failure paths — CLI-absent / empty / garbage stdin / unwritable tmp all exit 0.
- [x] PreToolUse matcher uses the plugin-scoped `mcp__plugin_meho_meho__call_operation` form; formula documented in the README.
- [x] Announce reminder at most once per session (marker dedupe) — `.reminded` gate; the Stop nudge is now separately bounded by `.stop_reminded`.
- [x] No blocking decision — grep of `hooks.json` + scripts shows no `deny` / `permissionDecision` / `exit 2`; warn/context output only.

### Project-board alignment
| Issue | Status | Link | Auto-close | Delivered | Action |
|---|---|---|---|---|---|
| #3131 | Open | Closes | ✅ | ✅ full | close on merge |
| #3128 | Open | parent-initiative | — | ⚠️ partial | stays open (wave 3) |
| #3130 | Closed | dependency | — | ✅ | none |
| #3133 | Closed | sibling (server gate) | — | ✅ | none |

### Predicted vs. actual diff
- ✅ Delta since iter-1 is exactly `bin/stop-report.sh` + `README.md` — the B1 fix scope, no creep.
- ✅ The other four files (`session-start.sh`, `pre-call-operation.sh`, `post-announce.sh`, `hooks.json`) are unchanged from iter-1 (which passed them).
- 🚫 none (no undisclosed edits; no `AGENTS.md` / `CLAUDE.md` / `.planning/` touched)
- 🕳️ none (all acceptance criteria delivered)

### Integration verification
- [x] `shellcheck -s sh bin/*.sh` — clean (all 4 scripts)
- [x] `jq -e . hooks/hooks.json` — valid; timeouts present
- [x] Stop-hook behavioral matrix (9 cases + 4 robustness cases) — all pass (remind-once, `stop_hook_active` bail, announced/unused/empty/garbage silent, cross-session isolation, traversal-safe)
- [x] File modes — all `bin/*.sh` remain `100755`
- [⚠️ n/a] `ruff` / `mypy` / `pytest` — no Python touched; shell + docs only. (The PR's IN_PROGRESS Python/Go/integration CI lanes run on every PR and cannot be affected by a shell/docs-only diff.)

### Findings
None. The sole iter-1 blocker (B1) is resolved; no new blockers, majors, or minors in the delta.

### Adjacent improvements (non-blocking, not posted inline)
Carried unchanged from iter-1 — recorded, not blocking merge:
- PreToolUse `additionalContext` lands next to the tool result (after the first `call_operation`), not before — inherent to the warn-only channel; worth a one-liner in `pre-call-operation.sh`.
- `session-start.sh` `status` call is unbounded when neither `timeout` nor `gtimeout` is on PATH (stock macOS); fail-open holds via the 12s hook timeout, but AC1's `<2s` no-VPN target isn't guaranteed there.
- Marker files under `${TMPDIR}/meho-plugin-hooks` are never swept (no SessionEnd cleanup) — benign but untidy.
- The `nosession` fallback bucket collides across concurrent sessions with an unparseable `session_id` — graceful-degradation edge case.
- No automated (bats) harness for the shell hooks — manual verification is documented and reproduced here; a bats harness would lock the fail-open/dedupe/`stop_hook_active` contract.

### Docs
- [x] `README.md` reflex-hooks section — accurate; Stop semantics corrected. No `docs/codebase/` doc applies (client plugin dir); `verified_unchanged` elsewhere.

### What worked well (review-body only)
- The B1 fix is minimal and surgical: two independent bounds (`stop_hook_active` + `.stop_reminded`) plus documentation truth-up, no scope creep.
- The `sed` extraction deliberately avoids BSD/GNU-divergent alternation and is false-positive-safe — a genuinely portable choice, verified across payload shapes.
